src/main/java/org/jenkinsci/plugins/workflow/steps/CoreWrapperStep.java

```diff
@@ -149,7 +213,53 @@ private void doStart() throws Exception {
         }
 
         @Override protected void finished(StepContext context) throws Exception {
-            disposer.tearDown(context.get(Run.class), context.get(FilePath.class), context.get(Launcher.class), context.get(TaskListener.class));
+            final Run<?,?> run = context.get(Run.class);
+            assert run != null;
+            final TaskListener listener = context.get(TaskListener.class);
+            assert listener != null;
+            final FilePath workspace = context.get(FilePath.class);
+            final Launcher launcher = context.get(Launcher.class);
+            boolean workspaceRequired = true;
+            // In Jenkins 2.25x, a Disposer has a final requiresWorkspace() method that indicates whether or not it (or
+            // more accurately, its associated wrapper) requires a workspace context. This is set up via the Context, as
+            // long as that is created via SimpleBuildWrapper.createContext().
+            // Note: this uses getMethod() on the disposer's type and not SimpleBuildWrapper.Disposer so that an
+            // implementation can get this behaviour without switching to Jenkins 2.25x itself.
+            // TODO: Use 'workspaceRequired = this.disposer.requiresWorkspace()' once this plugin depends on Jenkins 2.25x or later.
+            try {
+                final Method requiresWorkspace = this.disposer.getClass().getMethod("requiresWorkspace");
+                workspaceRequired = (boolean) requiresWorkspace.invoke(this.disposer);
+            } catch(NoSuchMethodException e) {
+                // ok, default to true
+            } catch (InvocationTargetException ite) {
+                final Throwable realException = ite.getCause();
+                throw realException instanceof Exception ? (Exception) realException : ite;
+            }
+            if (workspaceRequired) {
+                if (workspace == null) {
+                    throw new MissingContextVariableException(FilePath.class);
+                }
+                if (launcher == null) {
+                    throw new MissingContextVariableException(Launcher.class);
+                }
+            }
+            // always pass the workspace context when available, even when it is not strictly required
+            if (workspace != null && launcher != null) {
+                this.disposer.tearDown(run, workspace, launcher, listener);
+            } else {
+                // If we get here, workspaceRequired is false and there is no workspace context. In that case, the
+                // overload of tearDown() introduced in Jenkins 2.25x MUST exist.
+                // Note: this uses getMethod() on the disposer's type and not SimpleBuildWrapper.Disposer so that an
+                // implementation can get this behaviour without switching to Jenkins 2.25x itself.
+                // TODO: Use 'this.disposer.tearDown(run, listener)' once the minimum core version for this plugin is 2.25x or newer.
```

Comment:
```suggestion
                // overload of tearDown() introduced in Jenkins 2.258 MUST exist.
                // Note: this uses getMethod() on the disposer's type and not SimpleBuildWrapper.Disposer so that an
                // implementation can get this behaviour without switching to Jenkins 2.258 itself.
                // TODO: Use 'this.disposer.tearDown(run, listener)' once the minimum core version for this plugin is 2.258 or newer.
```